### PR TITLE
[WIP] Make all matrix properties accessible as @property

### LINKED ISSUE #15883: Make all matrix properties as @property
The properties in `MatrixProperties` class in `common.py` are not accessed in consistent manner right now.

| Properties | Functions |
|---|---|
| is_hermitian | is_symmetric |
| is_Identity | is_anti_symmetric |
| is_square | is_diagonal |
| is_lower_hessenberg | |
| is_upper_hessenberg | |
| is_lower | |
| is_upper | |

And also,

| Properties | Functions |
|---|---|
| free_symbols | atoms |
| | values |
| | is_symbolic |

These may not have been consistent,
though `has` may be leaved as-is like before.

One of the problem it had introduced, it that some optional arguments like `simplify`would not work once it is set to property,
```
from sympy import *
a = Matrix.eye(10)
a.is_hermitian(simplify=False)
```
unlike the code advertises like it should be
https://github.com/sympy/sympy/blob/cb8a524e930266fe2f4cda39d7d0b1523c4fa38b/sympy/matrices/common.py#L1155-L1156
because `a.is_hermitian` as a whole becomes a boolean object after called if set as property, and it would not be callable.
I think there would not be a possible way to override such thing in python nature, as something like `a.is_hermitian` to be used as property while also it would be accessed as `a.is_hermitian()` or `a.is_hermitian(simplify=False)`,
so literally it would be no different than making the function hardcoded with simplify.

It would also mean that it will be tricky to make compatible with legacies, if we make any changes.
Because it will simply not be usable as functions once we add `@property` decorator with it.

If it should be made changes with, I think there should be some compatibility layer added with some global variables, to make an option to choose whether to use Old_MatrixProperty class or New_MatrixProperty class,
so that it would be possible to throw some deprecation warning that you should change some global variable, and then it may be possible to use `a.is_symmetric` instead of `a.is_symmetric()` or such.

### LINKED ISSUE #16163: Deprecate _cache_is_diagonalizable and _cache_eigenvects
See issue #15887

Also remove these lines
```
    # XXX Remove these after deprecation is done for issue #15887
    '_cache_eigenvects', '_cache_is_diagonalizable']
```
in test_pickling

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Fixes #15883 
Fixes #16163

#### Brief description of what is fixed or changed

1. Make all matrix properties as `@property`
2. Remove all the keywords in the predicates
3. Use fuzzy logic.
  e.g. `Matrix([[a, b], [c, d]]).is_symmetric` should be `None` instead of `False`
4. Rename `is_zero` to `is_zero_matrix`

I'm not intending to change `is_diagonalizable_with_eigen` here, and I guess it can be changed to some other private method and the keyword can be kept.

Also I have temporarily removed the cacheing for `is_diagonalizable`, before I resolve test failings.

This change can be backward-incompatible, but as I have noted in #16163, raising deprecation warnings would be tricky here. (Or there can be some other possible suggestions).
But it's an obvious problem that things like `is_symmetric` and `is_hermitian` are organized, so at some point of release, this should be done.

I don't think that there should be a need to add keywords like `simplify` for computing assumptions, as I see there are no core assumptions handled like this way. 
I would first say that a lot of false-positives can be resolved if 3-valued logic is properly used.
And for simplifying, we should encourage expanding the assumptions into a full formal logic instead of this way.
For example, matrix symmetry can easily be made into a formal logic like image and if the logical expression stays unevaluated, it would only contain the specific indeterminate pairs, such that `simplify` can be expected to reduce.

#### Other comments

I may also try to split the properties into a different file for better organization.
I would wrap everything up in `MatrixProperties` or a new mixin class `MatrixAdvancedProperties` for the stuff like positive-definiteness, echelon, diagonalizability.

I've removed the test `test_issue_17247_expression_blowup_31` because the matrix looks like a non-symmetric matrix which LDL is not supposed to work (Though I'm not sure that why had it passed the test before)

#### Release Notes

- matrices
  - Made `is_symmetric`, `is_anti_symmetric`, `is_diagonal`, `is_symbolic`, `is_diagonalizable` and `is_nilpotent` accessible as `@property`
  - `is_nilpotent` will not raise error for non-square matrices, but return `False`.
  - Renamed `Matrix.is_zero` to `Matrix.is_zero_matrix`.
  - Made matrix properties return fuzzy values if it is logically undetermined. e.g. `Matrix([[a, b], [c, d]]).is_symmetric` will return `None` instead of `False`.
- integrals
  - Fixed integrals returning zero instead of zero matrix for integrating zero matrices. e.g. `integrate(Matrix([[0, 0], [0, 0]]), x)`

### CODE REVIEW

**Comment by asmeurer** [comment]
I don't think there's a way to do a deprecation here. It would help to list out all the attributes that were changed in the release notes.

**Comment by asmeurer** [comment]
It should be noted that properties remove the possibility in the future of adding flags to the methods.

**Comment by sylee957** [comment]
Cyclic import problem is making it difficult to separate `MatrixProperties` to a separate file

File: sympy/matrices/sparse.py [hunk 1]
```diff
@@ -463,15 +463,6 @@ def _eval_has(self, *patterns):
...
         return any(self[key].has(*patterns) for key in self._smat) or zhas
 
-    def _eval_is_Identity(self):
-        if not all(self[i, i] == 1 for i in range(self.rows)):
-            return False
-        return len(self._smat) == self.rows
-
-    def _eval_is_symmetric(self, simpfunc):
-        diff = (self - self.T).applyfunc(simpfunc)
-        return len(diff.values()) == 0
```
Comment (sylee957) [inline sympy/matrices/sparse.py]: The judgements from sparsity alone can make false-positives for symbolic matrices, so I have removed them for now.

**Comment by sylee957** [comment]
I'll also take a look at how integral is working for https://github.com/sympy/sympy/issues/15883#issuecomment-570964395

**Comment by oscarbenjamin** [comment]
I think this looks good.

Clearly this is a backward incompatible change though and I'm not sure how to gauge the impact of it.

The release notes should explicitly list each method that is changed to a property with a clear statement that explicitly mentions backward compatibility.

**Comment by oscarbenjamin** [comment]
I like all the changes here but some are not backwards-compatible. I understand why you want to break with backwards compatibility but can you explain the likely impact on users?

**Comment by sylee957** [comment]
Okay, there are two ways that it can break the code
First would be every code using it would raise `TypeError`
```python3
if M.is_symmetric():
```
And second would be every code expecting a keyword argument
```python3
if M.is_symmetric(simplify=True):
```
would also raise `TypeError`.

And also it may deprecate the variation of `is_diagonalizable` like `reals_only`. 
But I would try to defend it that the previous notation can be reduced to `all(x.is_real for x in eigenvals)`

**Comment by asmeurer** [comment]
Basically any user code that uses one of these attributes will break. A change like this shouldn't be made lightly.

**Comment by sylee957** [comment]
If it is not merged, I may open up a separate PR regarding the changes to the logic.
Let’s wait some more if this gets more vote.

**Comment by sylee957** [comment]
On my second thought, I suppose that if we are facing things that are just unchangable without much code breaks, it would be a better idea to switch on and introduce a matrix version 2 API with a different name, with all those weird designs resolved.
And we would keep the old matrix only for some maintenance.

**Comment by Pristine-Cat** [comment]
Are you really thinking of introducing a separate matrix class? That could resolve many things...

**Comment by oscarbenjamin** [comment]
> Are you really thinking of introducing a separate matrix class? That could resolve many things...

I think it could. That's part of why I think that moving non-trivial code out of the matrix classes themselves would be helpful.

**Comment by Pristine-Cat** [comment]
We could maybe even call the new matrix classes ... `IMatrix` and `MMatrix`??? :) :) :) And if you want to move that code out faster then lets finish the determinant stuff so I can pull out the other stuff as well.

**Comment by sylee957** [comment]
Yes, similar stuff can be done for some hard-to-resolve backward incompatible changes in #18088
But I would also learn from this that bad designs really cost a lot.

Personally, to make good designs for a lot of stuff like how to organize the property, I think that it was some greed from the people in the past, trying to implement so many 'controls'.

**Comment by sylee957** [comment]
> We could maybe even call the new matrix classes ... `IMatrix` and `MMatrix`??? :) :) :) And if you want to move that code out faster then lets finish the determinant stuff so I can pull out the other stuff as well (am waiting on your decision in that branch).

Yes, `IMatrix` and `MMatrix` can be a name for a matrix V2 classes, but I would try to make every matrix properties symmetrically accessible as `@property` there.
I would also have a lot of words to say for deprecating some other obscure APIs like `LUdecomposition_Simple` and renaming some stuff.

**Comment by oscarbenjamin** [comment]
> Yes, `IMatrix` and `MMatrix` can be a name for a matrix V2 classes

We need a clearly fleshed out design before implementing anything

**Comment by czgdp1807** [comment]
@sylee957 Please resolve the conflicts.

**Comment by sylee957** [comment]
I'd defer this to be discussed in the matrix refactories #18445
I guess it already looks too messy and sympy architecture does not provide much tools to manage the assumption predicates in one place.

**Pull request closed without merging**